<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-99.9.9">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Notes - QC and Cell-level filtering</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Notes</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/mmyoung/notebook-241114"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text">Source code</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#normalizing-the-data" id="toc-normalizing-the-data" class="nav-link active" data-scroll-target="#normalizing-the-data">Normalizing the data</a></li>
  <li><a href="#cell-clustering" id="toc-cell-clustering" class="nav-link" data-scroll-target="#cell-clustering">cell clustering</a></li>
  <li><a href="#define-doublet" id="toc-define-doublet" class="nav-link" data-scroll-target="#define-doublet">define doublet</a>
  <ul class="collapse">
  <li><a href="#reclustering-on-the-filtered-cells" id="toc-reclustering-on-the-filtered-cells" class="nav-link" data-scroll-target="#reclustering-on-the-filtered-cells">reclustering on the filtered cells</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">QC and Cell-level filtering</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div id="c918651d-4fa0-4afc-9632-787aee36c0e5" class="cell" data-scrolled="true" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>library(Seurat)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>library(ggplot2)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>library(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>library(patchwork)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>library(stringr)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>library(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="3a15457c-745f-4a9b-a469-ef30a0a747e5" class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">## load gene by cell matrices from all samples into individual seurat objects</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>seurat_obj_ls <span class="op">&lt;-</span> </span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    lapply(c(<span class="st">'iso-F-04'</span>,<span class="st">'iso-F-22'</span>,<span class="st">'iso-F-35'</span>,<span class="st">'iso-M-00'</span>,<span class="st">'iso-M-23'</span>,<span class="st">'iso-M-40'</span>,<span class="st">'soc-F-05'</span>,<span class="st">'soc-F-24'</span>,<span class="st">'soc-F-36'</span>,<span class="st">'soc-M-01'</span>,<span class="st">'soc-M-25'</span>,<span class="st">'soc-M-41'</span>),</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>       function(x){</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>           cell.use <span class="op">&lt;-</span> read.csv(paste0(<span class="st">"./samples/"</span>,x,<span class="st">".Rat.allCells.csv"</span>))<span class="op">%&gt;%</span> </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>            <span class="bu">filter</span>(<span class="cf">pass</span><span class="op">==</span><span class="st">"True"</span>) <span class="op">%&gt;%</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>            pull(X)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>        Read10X(paste0(<span class="st">"alignment/"</span>,x,<span class="st">".Rat/"</span>,x,<span class="st">".Rat.star.solo/GeneFull_Ex50pAS/raw/"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>           CreateSeuratObject(project <span class="op">=</span> <span class="st">"Rat"</span>, <span class="bu">min</span>.cells <span class="op">=</span> <span class="dv">3</span>, <span class="bu">min</span>.features <span class="op">=</span> <span class="dv">200</span>) <span class="op">%&gt;%</span>       </span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>            subset(cells <span class="op">=</span> cell.use)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>           })</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="a1c13834-c892-4e19-934f-14fce87b360a" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">## merge all individual objects into one</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>Seurat_merged <span class="op">&lt;-</span> merge(seurat_obj_ls[[<span class="dv">1</span>]],seurat_obj_ls[<span class="op">-</span><span class="dv">1</span>],project<span class="op">=</span><span class="st">"Rat"</span>,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>                       add.cell.ids <span class="op">=</span>c(<span class="st">'iso-F-04'</span>,<span class="st">'iso-F-22'</span>,<span class="st">'iso-F-35'</span>,<span class="st">'iso-M-00'</span>,<span class="st">'iso-M-23'</span>,<span class="st">'iso-M-40'</span>,<span class="st">'soc-F-05'</span>,<span class="st">'soc-F-24'</span>,<span class="st">'soc-F-36'</span>,<span class="st">'soc-M-01'</span>,<span class="st">'soc-M-25'</span>,<span class="st">'soc-M-41'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="f18ba525" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="4">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>Seurat_merged</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An object of class Seurat 
26597 features across 76385 samples within 1 assay 
Active assay: RNA (26597 features, 0 variable features)
 12 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, counts.8, counts.9, counts.10, counts.11, counts.12</code></pre>
</div>
</div>
<div id="9fefede4-e5be-4b0f-9e59-21f68414281f" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>Seurat_merged[[<span class="st">"RNA"</span>]] <span class="op">&lt;-</span>JoinLayers(Seurat_merged[[<span class="st">"RNA"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="64f29b13-50e8-4b58-991f-2f405eceb9c3" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>Seurat_merged</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An object of class Seurat 
26597 features across 76385 samples within 1 assay 
Active assay: RNA (26597 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
</div>
<div id="09009487-aeaa-4e02-8da9-f69a9de119fa" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="7">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The [[ operator can add columns to object metadata. This is a great place to stash QC stats</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>Seurat_merged[[<span class="st">"percent.mt"</span>]] <span class="op">&lt;-</span> PercentageFeatureSet(Seurat_merged, pattern <span class="op">=</span> <span class="st">"^Mt-"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="7918784d-b17e-4717-a561-e03ab358b6cb" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="8">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">10</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">5</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cf03d99a-deb1-4c6d-a79e-486b8c1b37fa" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="9">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize QC metrics as a violin plot</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="op">&lt;-</span> VlnPlot(Seurat_merged, features <span class="op">=</span> c(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>), ncol <span class="op">=</span> <span class="dv">3</span>,layer<span class="op">=</span><span class="st">'counts'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="62b79681-0b5e-4b12-9ba5-69f24b185c03" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>p1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-11-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="18f85702-c5bd-4346-8c8e-008a39caba07" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="11">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">## filter cells with detected gene number between 200 and 4000</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>obj <span class="op">&lt;-</span> subset(Seurat_merged, subset <span class="op">=</span> nFeature_RNA <span class="op">&gt;</span> <span class="dv">200</span> <span class="op">&amp;</span> nFeature_RNA <span class="op">&lt;</span> <span class="dv">4000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="22ba08c2-ed21-4d97-be1b-7076164f6930" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>obj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An object of class Seurat 
26597 features across 76089 samples within 1 assay 
Active assay: RNA (26597 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
</div>
<section id="normalizing-the-data" class="level2">
<h2 class="anchored" data-anchor-id="normalizing-the-data">Normalizing the data</h2>
<div id="f2c233ef-d34b-4d4a-a200-d1a9eb9d8c9c" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="13">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">## using SCTransform to normalize the gene counts</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>obj <span class="op">&lt;-</span> SCTransform(obj, vst.flavor <span class="op">=</span> <span class="st">'v2'</span>, verbose <span class="op">=</span> FALSE) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="340f325c-8f89-4110-a66c-61551c39eb21" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="14">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Run PCA dimention reduction on the normalized counts</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>obj <span class="op">&lt;-</span> obj <span class="op">%&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    RunPCA(npcs <span class="op">=</span> <span class="dv">30</span>, verbose <span class="op">=</span> FALSE,features <span class="op">=</span> VariableFeatures(<span class="bu">object</span> <span class="op">=</span> obj))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="4997d177-8f8e-4836-96b7-58fcfd9a5386" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="15">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>VizDimLoadings(obj, dims <span class="op">=</span> <span class="dv">1</span>:<span class="dv">2</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-16-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="a843411f-9e64-4faf-818a-66a6fbaa11ca" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="16">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">10</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">8</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>DimHeatmap(obj, dims <span class="op">=</span> <span class="dv">1</span>:<span class="dv">9</span>, cells <span class="op">=</span> <span class="dv">500</span>, balanced <span class="op">=</span> TRUE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-17-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="935df065-783d-4527-bf68-16be85cdcc34" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="17">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">5</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">4</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>ElbowPlot(obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-18-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="cell-clustering" class="level2">
<h2 class="anchored" data-anchor-id="cell-clustering">cell clustering</h2>
<div id="081375eb-319d-4c6b-b75a-82aa5167d8da" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>obj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An object of class Seurat 
52115 features across 76089 samples within 2 assays 
Active assay: SCT (25518 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 1 dimensional reduction calculated: pca</code></pre>
</div>
</div>
<div id="6f31334b-362b-4353-889e-2e9fe064ff95" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">## add the group and sex information of each cell in the metadata</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>obj <span class="op">&lt;-</span> AddMetaData(</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">object</span> <span class="op">=</span> obj,</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  metadata <span class="op">=</span> stringr::str_split(row.names(obj<span class="op">@</span>meta.data),<span class="st">"-"</span>,simplify<span class="op">=</span>T)[,<span class="dv">1</span>],</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  col.name <span class="op">=</span> <span class="st">'group'</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>obj <span class="op">&lt;-</span> AddMetaData(</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  <span class="bu">object</span> <span class="op">=</span> obj,</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>  metadata <span class="op">=</span> stringr::str_split(row.names(obj<span class="op">@</span>meta.data),<span class="st">"-"</span>,simplify<span class="op">=</span>T)[,<span class="dv">2</span>],</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>  col.name <span class="op">=</span> <span class="st">'sex'</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="4d9f614f-e9f4-4fd4-af38-e73dd7217c7e" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>head(obj<span class="op">@</span>meta.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 8</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">orig.ident</th>
<th data-quarto-table-cell-role="th" scope="col">nCount_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">percent.mt</th>
<th data-quarto-table-cell-role="th" scope="col">nCount_SCT</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_SCT</th>
<th data-quarto-table-cell-role="th" scope="col">group</th>
<th data-quarto-table-cell-role="th" scope="col">sex</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGACTATGCAATACCAGGCGCA</td>
<td>Rat</td>
<td>2892</td>
<td>1519</td>
<td>0</td>
<td>3229</td>
<td>1519</td>
<td>iso</td>
<td>F</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGACTATGCAATTCGGATTCGG</td>
<td>Rat</td>
<td>2971</td>
<td>1489</td>
<td>0</td>
<td>3287</td>
<td>1489</td>
<td>iso</td>
<td>F</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGGGTATGCTTGTCGGATTCGG</td>
<td>Rat</td>
<td>1996</td>
<td>1189</td>
<td>0</td>
<td>3028</td>
<td>1208</td>
<td>iso</td>
<td>F</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCCTACCAGTCTAACTAGGT</td>
<td>Rat</td>
<td>5871</td>
<td>2496</td>
<td>0</td>
<td>4299</td>
<td>2456</td>
<td>iso</td>
<td>F</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCCTACCAGTGTAGGATAAG</td>
<td>Rat</td>
<td>2068</td>
<td>1185</td>
<td>0</td>
<td>3104</td>
<td>1190</td>
<td>iso</td>
<td>F</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCGCGGTTGGTTGATACGAT</td>
<td>Rat</td>
<td>1922</td>
<td>1121</td>
<td>0</td>
<td>3064</td>
<td>1153</td>
<td>iso</td>
<td>F</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="d9428cd5-4385-491d-b259-74143285b4f5" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">## run cell clustering with the first 15 PCs</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>obj <span class="op">&lt;-</span> obj <span class="op">%&gt;%</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>        FindNeighbors(reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="dv">1</span>:<span class="dv">15</span>, verbose <span class="op">=</span> FALSE) <span class="op">%&gt;%</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>        FindClusters(resolution <span class="op">=</span> <span class="fl">0.7</span>, verbose <span class="op">=</span> FALSE) <span class="op">%&gt;%</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>        RunUMAP(reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="dv">1</span>:<span class="dv">15</span>, verbose <span class="op">=</span> FALSE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="625b1c9f-26f2-49f1-8d6c-0c457cf3c3d2" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>obj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An object of class Seurat 
52115 features across 76089 samples within 2 assays 
Active assay: SCT (25518 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: pca, umap</code></pre>
</div>
</div>
<div id="876102c6-6144-481b-bdfc-c6c1d77df035" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="23">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co">## umap plot</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">7</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">5</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="co"># note that you can set `label = TRUE` or use the LabelClusters function to help label</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="co"># individual clusters</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>DimPlot(obj, reduction <span class="op">=</span> <span class="st">"umap"</span>,label <span class="op">=</span> TRUE, repel <span class="op">=</span> TRUE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-24-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="define-doublet" class="level2">
<h2 class="anchored" data-anchor-id="define-doublet">define doublet</h2>
<p>remove the doublets with DoubleFinder</p>
<div id="23985340-8bed-44bf-8ee0-071345070034" class="cell" data-scrolled="true">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>library(<span class="st">'DoubletFinder'</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>homotypic.prop <span class="op">&lt;-</span> modelHomotypic(obj<span class="op">@</span>meta.data$seurat_clusters)           <span class="co">## ex: annotations &lt;- seu_kidney@meta.data$ClusteringResults</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>nExp_poi <span class="op">&lt;-</span> <span class="bu">round</span>(<span class="fl">0.075</span><span class="op">*</span>nrow(obj<span class="op">@</span>meta.data))  <span class="co">## Assuming 7.5% doublet formation rate - tailor for your dataset</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>nExp_poi.adj <span class="op">&lt;-</span> <span class="bu">round</span>(nExp_poi<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>homotypic.prop))</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="co">## Run DoubletFinder with varying classification stringencies ----------------------------------------------------------------</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>seu_obj <span class="op">&lt;-</span> doubletFinder(obj, PCs <span class="op">=</span> <span class="dv">1</span>:<span class="dv">10</span>, pN <span class="op">=</span> <span class="fl">0.25</span>, pK <span class="op">=</span> <span class="fl">0.09</span>, nExp <span class="op">=</span> nExp_poi, reuse.pANN <span class="op">=</span> FALSE, sct <span class="op">=</span> TRUE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="08d38420-c73b-4a69-939b-703760dd9cf9" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>head(seu_obj<span class="op">@</span>meta.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 12</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">orig.ident</th>
<th data-quarto-table-cell-role="th" scope="col">nCount_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">percent.mt</th>
<th data-quarto-table-cell-role="th" scope="col">nCount_SCT</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_SCT</th>
<th data-quarto-table-cell-role="th" scope="col">group</th>
<th data-quarto-table-cell-role="th" scope="col">sex</th>
<th data-quarto-table-cell-role="th" scope="col">SCT_snn_res.0.7</th>
<th data-quarto-table-cell-role="th" scope="col">seurat_clusters</th>
<th data-quarto-table-cell-role="th" scope="col">pANN_0.25_0.09_5707</th>
<th data-quarto-table-cell-role="th" scope="col">DF.classifications_0.25_0.09_5707</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGACTATGCAATACCAGGCGCA</td>
<td>Rat</td>
<td>2892</td>
<td>1519</td>
<td>0</td>
<td>3229</td>
<td>1519</td>
<td>iso</td>
<td>F</td>
<td>9</td>
<td>9</td>
<td>0.3220896</td>
<td>Singlet</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGACTATGCAATTCGGATTCGG</td>
<td>Rat</td>
<td>2971</td>
<td>1489</td>
<td>0</td>
<td>3287</td>
<td>1489</td>
<td>iso</td>
<td>F</td>
<td>10</td>
<td>10</td>
<td>0.2795970</td>
<td>Singlet</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGGGTATGCTTGTCGGATTCGG</td>
<td>Rat</td>
<td>1996</td>
<td>1189</td>
<td>0</td>
<td>3028</td>
<td>1208</td>
<td>iso</td>
<td>F</td>
<td>9</td>
<td>9</td>
<td>0.3189136</td>
<td>Singlet</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCCTACCAGTCTAACTAGGT</td>
<td>Rat</td>
<td>5871</td>
<td>2496</td>
<td>0</td>
<td>4299</td>
<td>2456</td>
<td>iso</td>
<td>F</td>
<td>14</td>
<td>14</td>
<td>0.2988720</td>
<td>Singlet</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCCTACCAGTGTAGGATAAG</td>
<td>Rat</td>
<td>2068</td>
<td>1185</td>
<td>0</td>
<td>3104</td>
<td>1190</td>
<td>iso</td>
<td>F</td>
<td>5</td>
<td>5</td>
<td>0.1328442</td>
<td>Singlet</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCGCGGTTGGTTGATACGAT</td>
<td>Rat</td>
<td>1922</td>
<td>1121</td>
<td>0</td>
<td>3064</td>
<td>1153</td>
<td>iso</td>
<td>F</td>
<td>9</td>
<td>9</td>
<td>0.3220896</td>
<td>Singlet</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="529ebab4-a3c7-44db-a7b9-5d3e6811d939" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">6</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">4</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co"># note that you can set `label = TRUE` or use the LabelClusters function to help label</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co"># individual clusters</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>DimPlot(seu_obj, reduction <span class="op">=</span> <span class="st">"umap"</span>,label <span class="op">=</span> TRUE, group.by <span class="op">=</span> <span class="st">"DF.classifications_0.25_0.09_5707"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-27-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="1ebcde5e-94a2-4d5c-8680-06f8c346e5e9" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>seu_obj <span class="op">&lt;-</span> subset(x <span class="op">=</span> seu_obj, subset <span class="op">=</span> DF.classifications_0<span class="fl">.25_0.09_5707</span><span class="op">==</span><span class="st">"Singlet"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="a9da8911-dc54-45c8-82a8-07cdea01755d" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>seu_obj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An object of class Seurat 
52115 features across 70382 samples within 2 assays 
Active assay: SCT (25518 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: pca, umap</code></pre>
</div>
</div>
<section id="reclustering-on-the-filtered-cells" class="level3">
<h3 class="anchored" data-anchor-id="reclustering-on-the-filtered-cells">reclustering on the filtered cells</h3>
<div id="02c87211-599a-4dc5-a7af-b05cba175261" class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>seu_obj <span class="op">&lt;-</span> seu_obj <span class="op">%&gt;%</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>        FindNeighbors(reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="dv">1</span>:<span class="dv">15</span>, verbose <span class="op">=</span> FALSE) <span class="op">%&gt;%</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>        FindClusters(resolution <span class="op">=</span> <span class="fl">0.7</span>, verbose <span class="op">=</span> FALSE) <span class="op">%&gt;%</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>        RunUMAP(reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="dv">1</span>:<span class="dv">15</span>, verbose <span class="op">=</span> FALSE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="70261b40-970a-46c6-9751-616aa1d3c13b" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">7</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">5</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="co"># note that you can set `label = TRUE` or use the LabelClusters function to help label</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="co"># individual clusters</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>DimPlot(seu_obj, reduction <span class="op">=</span> <span class="st">"umap"</span>,label <span class="op">=</span> TRUE, repel <span class="op">=</span> TRUE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-31-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="e9b16d8c-f271-4b7b-94c7-75b069a01353" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>library(dplyr)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>library(stringr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="7c5ea7f2-7091-4770-9dce-ea92fbe4f4b3" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>head(seu_obj<span class="op">@</span>meta.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 12</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">orig.ident</th>
<th data-quarto-table-cell-role="th" scope="col">nCount_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">percent.mt</th>
<th data-quarto-table-cell-role="th" scope="col">nCount_SCT</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_SCT</th>
<th data-quarto-table-cell-role="th" scope="col">group</th>
<th data-quarto-table-cell-role="th" scope="col">sex</th>
<th data-quarto-table-cell-role="th" scope="col">SCT_snn_res.0.7</th>
<th data-quarto-table-cell-role="th" scope="col">seurat_clusters</th>
<th data-quarto-table-cell-role="th" scope="col">pANN_0.25_0.09_5707</th>
<th data-quarto-table-cell-role="th" scope="col">DF.classifications_0.25_0.09_5707</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGACTATGCAATACCAGGCGCA</td>
<td>Rat</td>
<td>2892</td>
<td>1519</td>
<td>0</td>
<td>3229</td>
<td>1519</td>
<td>iso</td>
<td>F</td>
<td>10</td>
<td>10</td>
<td>0.3220896</td>
<td>Singlet</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGACTATGCAATTCGGATTCGG</td>
<td>Rat</td>
<td>2971</td>
<td>1489</td>
<td>0</td>
<td>3287</td>
<td>1489</td>
<td>iso</td>
<td>F</td>
<td>8</td>
<td>8</td>
<td>0.2795970</td>
<td>Singlet</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGGGTATGCTTGTCGGATTCGG</td>
<td>Rat</td>
<td>1996</td>
<td>1189</td>
<td>0</td>
<td>3028</td>
<td>1208</td>
<td>iso</td>
<td>F</td>
<td>10</td>
<td>10</td>
<td>0.3189136</td>
<td>Singlet</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCCTACCAGTCTAACTAGGT</td>
<td>Rat</td>
<td>5871</td>
<td>2496</td>
<td>0</td>
<td>4299</td>
<td>2456</td>
<td>iso</td>
<td>F</td>
<td>14</td>
<td>14</td>
<td>0.2988720</td>
<td>Singlet</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCCTACCAGTGTAGGATAAG</td>
<td>Rat</td>
<td>2068</td>
<td>1185</td>
<td>0</td>
<td>3104</td>
<td>1190</td>
<td>iso</td>
<td>F</td>
<td>3</td>
<td>3</td>
<td>0.1328442</td>
<td>Singlet</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCGCGGTTGGTTGATACGAT</td>
<td>Rat</td>
<td>1922</td>
<td>1121</td>
<td>0</td>
<td>3064</td>
<td>1153</td>
<td>iso</td>
<td>F</td>
<td>10</td>
<td>10</td>
<td>0.3220896</td>
<td>Singlet</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Add substitutional sample name to metadata</p>
<div id="943b1c92-3c07-44e4-ad9d-da6da99e8ae7" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>seu_obj<span class="op">@</span>meta.data$sampleID <span class="op">&lt;-</span> str_split(row.names(seu_obj<span class="op">@</span>meta.data),<span class="st">"_"</span>,simplify <span class="op">=</span> T)[,<span class="dv">1</span>]</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>seu_obj<span class="op">@</span>meta.data <span class="op">&lt;-</span> </span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>    seu_obj<span class="op">@</span>meta.data <span class="op">%&gt;%</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>    mutate(sample_name <span class="op">=</span> case_when(sampleID <span class="op">==</span> <span class="st">"iso-F-04"</span> <span class="op">~</span> <span class="st">"iso-F_1"</span>,</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>                                   sampleID <span class="op">==</span> <span class="st">"iso-F-22"</span> <span class="op">~</span> <span class="st">"iso-F_2"</span>,</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>                                   sampleID <span class="op">==</span> <span class="st">"iso-F-35"</span> <span class="op">~</span> <span class="st">"iso-F_3"</span>,</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>                                   sampleID <span class="op">==</span> <span class="st">"iso-M-00"</span> <span class="op">~</span> <span class="st">"iso-M_1"</span>,</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>                                   sampleID <span class="op">==</span> <span class="st">"iso-M-23"</span> <span class="op">~</span> <span class="st">"iso-M_2"</span>,</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>                                   sampleID <span class="op">==</span> <span class="st">"iso-M-40"</span> <span class="op">~</span> <span class="st">"iso-M_3"</span>,</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>                                   sampleID <span class="op">==</span> <span class="st">"soc-F-05"</span> <span class="op">~</span> <span class="st">"soc-F_1"</span>,</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>                                   sampleID <span class="op">==</span> <span class="st">"soc-F-24"</span> <span class="op">~</span> <span class="st">"soc-F_2"</span>,</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>                                   sampleID <span class="op">==</span> <span class="st">"soc-F-36"</span> <span class="op">~</span> <span class="st">"soc-F_3"</span>,</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>                                   sampleID <span class="op">==</span> <span class="st">"soc-M-01"</span> <span class="op">~</span> <span class="st">"soc-M_1"</span>,</span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>                                   sampleID <span class="op">==</span> <span class="st">"soc-M-25"</span> <span class="op">~</span> <span class="st">"soc-M_2"</span>,</span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>                                   sampleID <span class="op">==</span> <span class="st">"soc-M-41"</span> <span class="op">~</span> <span class="st">"soc-M_3"</span></span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>                                  ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="a061369d-eebc-4b6a-81cc-cba4f2478366" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">10</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">5</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>DimPlot(seu_obj, reduction <span class="op">=</span> <span class="st">"umap"</span>, split.by <span class="op">=</span> <span class="st">"sample_name"</span>,alpha<span class="op">=</span><span class="fl">0.7</span>,ncol<span class="op">=</span><span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-35-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="5e4aec4f-d797-4f3b-bf0e-359b8103dd83" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>saveRDS(seu_obj,<span class="bu">file</span><span class="op">=</span><span class="st">"/project/zhuzhuzhang/lyang/scaleBio_RNA_rat_brain/out_mRatBN/merged_seurat_obj_DF_removed_2.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="09b4af26-5a77-4fca-8164-5210535f663e" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>head(seu_obj<span class="op">@</span>meta.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 14</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">orig.ident</th>
<th data-quarto-table-cell-role="th" scope="col">nCount_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">percent.mt</th>
<th data-quarto-table-cell-role="th" scope="col">nCount_SCT</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_SCT</th>
<th data-quarto-table-cell-role="th" scope="col">group</th>
<th data-quarto-table-cell-role="th" scope="col">sex</th>
<th data-quarto-table-cell-role="th" scope="col">SCT_snn_res.0.7</th>
<th data-quarto-table-cell-role="th" scope="col">seurat_clusters</th>
<th data-quarto-table-cell-role="th" scope="col">pANN_0.25_0.09_5707</th>
<th data-quarto-table-cell-role="th" scope="col">DF.classifications_0.25_0.09_5707</th>
<th data-quarto-table-cell-role="th" scope="col">sampleID</th>
<th data-quarto-table-cell-role="th" scope="col">sample_name</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGACTATGCAATACCAGGCGCA</td>
<td>Rat</td>
<td>2892</td>
<td>1519</td>
<td>0</td>
<td>3229</td>
<td>1519</td>
<td>iso</td>
<td>F</td>
<td>10</td>
<td>10</td>
<td>0.3220896</td>
<td>Singlet</td>
<td>iso-F-04</td>
<td>iso-F_1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGACTATGCAATTCGGATTCGG</td>
<td>Rat</td>
<td>2971</td>
<td>1489</td>
<td>0</td>
<td>3287</td>
<td>1489</td>
<td>iso</td>
<td>F</td>
<td>8</td>
<td>8</td>
<td>0.2795970</td>
<td>Singlet</td>
<td>iso-F-04</td>
<td>iso-F_1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGGGTATGCTTGTCGGATTCGG</td>
<td>Rat</td>
<td>1996</td>
<td>1189</td>
<td>0</td>
<td>3028</td>
<td>1208</td>
<td>iso</td>
<td>F</td>
<td>10</td>
<td>10</td>
<td>0.3189136</td>
<td>Singlet</td>
<td>iso-F-04</td>
<td>iso-F_1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCCTACCAGTCTAACTAGGT</td>
<td>Rat</td>
<td>5871</td>
<td>2496</td>
<td>0</td>
<td>4299</td>
<td>2456</td>
<td>iso</td>
<td>F</td>
<td>14</td>
<td>14</td>
<td>0.2988720</td>
<td>Singlet</td>
<td>iso-F-04</td>
<td>iso-F_1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCCTACCAGTGTAGGATAAG</td>
<td>Rat</td>
<td>2068</td>
<td>1185</td>
<td>0</td>
<td>3104</td>
<td>1190</td>
<td>iso</td>
<td>F</td>
<td>3</td>
<td>3</td>
<td>0.1328442</td>
<td>Singlet</td>
<td>iso-F-04</td>
<td>iso-F_1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">iso-F-04_AAACCATAGTCGCGGTTGGTTGATACGAT</td>
<td>Rat</td>
<td>1922</td>
<td>1121</td>
<td>0</td>
<td>3064</td>
<td>1153</td>
<td>iso</td>
<td>F</td>
<td>10</td>
<td>10</td>
<td>0.3220896</td>
<td>Singlet</td>
<td>iso-F-04</td>
<td>iso-F_1</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="78fb5417-01f9-45be-a87c-be8603f78d71" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">5</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">3</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>seu_obj<span class="op">@</span>meta.data <span class="op">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    ggplot(aes(x<span class="op">=</span>sample_name))<span class="op">+</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>    geom_bar(stat<span class="op">=</span><span class="st">"count"</span>)<span class="op">+</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>    theme_minimal()<span class="op">+</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>    ggtitle(<span class="st">"Cell numbers in sample"</span>)<span class="op">+</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>    theme(axis.text.x <span class="op">=</span> element_text(angle<span class="op">=</span><span class="dv">30</span>,vjust<span class="op">=</span><span class="dv">1</span>,hjust<span class="op">=</span><span class="dv">1</span>))<span class="op">+</span></span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>    labs(x<span class="op">=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-38-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="389497f9-5afc-4e6b-b995-d8d7983f62fe" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">5</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">3</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>seu_obj<span class="op">@</span>meta.data <span class="op">%&gt;%</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>    ggplot(aes(x<span class="op">=</span>sample_name,y<span class="op">=</span>nFeature_RNA))<span class="op">+</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>    geom_violin(width<span class="op">=</span><span class="dv">1</span>,position <span class="op">=</span> position_dodge(<span class="fl">0.9</span>)) <span class="op">+</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>    geom_boxplot(width<span class="op">=</span><span class="fl">0.2</span>, color<span class="op">=</span><span class="st">"grey"</span>, alpha<span class="op">=</span><span class="fl">0.4</span>,position <span class="op">=</span> position_dodge(<span class="fl">0.9</span>))<span class="op">+</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>    theme_minimal()<span class="op">+</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>    ggtitle(<span class="st">"gene counts in each cell"</span>)<span class="op">+</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>    theme(axis.text.x <span class="op">=</span> element_text(angle<span class="op">=</span><span class="dv">30</span>,vjust<span class="op">=</span><span class="dv">1</span>,hjust<span class="op">=</span><span class="dv">1</span>))<span class="op">+</span></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    labs(x<span class="op">=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-39-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="968a8083-db98-4832-8919-a64f5b98b6db" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">10</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">5</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="op">&lt;-</span> DimPlot(seu_obj, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"group"</span>,alpha<span class="op">=</span><span class="fl">0.6</span>)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>p2 <span class="op">&lt;-</span> DimPlot(seu_obj, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"sex"</span>,alpha<span class="op">=</span><span class="fl">0.7</span>)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>p1<span class="op">+</span>p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-40-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="d0eda20b-4533-475c-8ffd-df09b9e229a8" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>seu_obj<span class="op">@</span>meta.data <span class="op">%&gt;%</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>    ggplot(aes(x<span class="op">=</span>seurat_clusters,y<span class="op">=</span>nCount_RNA))<span class="op">+</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>    geom_boxplot()<span class="op">+</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>    theme_minimal()<span class="op">+</span></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>    scale_x_discrete(name<span class="op">=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="cell_clustering_files/figure-html/cell-41-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    const typesetMath = (el) => {
      if (window.MathJax) {
        // MathJax Typeset
        window.MathJax.typeset([el]);
      } else if (window.katex) {
        // KaTeX Render
        var mathElements = el.getElementsByClassName("math");
        var macros = [];
        for (var i = 0; i < mathElements.length; i++) {
          var texText = mathElements[i].firstChild;
          if (mathElements[i].tagName == "SPAN") {
            window.katex.render(texText.data, mathElements[i], {
              displayMode: mathElements[i].classList.contains('display'),
              throwOnError: false,
              macros: macros,
              fleqn: false
            });
          }
        }
      }
    }
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        for (let i = 0; i < 2; i++) {
          container.appendChild(note.children[i].cloneNode(true));
        }
        typesetMath(container);
        return container.innerHTML
      } else {
        typesetMath(note);
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      typesetMath(note);
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>