<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-99.9.9">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Notes - plot for the DEG number in each cell type</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Notes</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/mmyoung/notebook-241114"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text">Source code</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#violin-plots-for-genes" id="toc-violin-plots-for-genes" class="nav-link active" data-scroll-target="#violin-plots-for-genes">violin plots for genes</a></li>
  <li><a href="#save-the-data-matrix-for-all-degs-between-f-and-m" id="toc-save-the-data-matrix-for-all-degs-between-f-and-m" class="nav-link" data-scroll-target="#save-the-data-matrix-for-all-degs-between-f-and-m">save the data matrix for all DEGs between F and M</a></li>
  <li><a href="#save-the-fc-data-for-all-celltypes-for-all-degs" id="toc-save-the-fc-data-for-all-celltypes-for-all-degs" class="nav-link" data-scroll-target="#save-the-fc-data-for-all-celltypes-for-all-degs">save the FC data for all celltypes for all DEGs</a>
  <ul class="collapse">
  <li><a href="#to-make-the-asterisk-labeled-heatmap-for-f_vs_m-degs" id="toc-to-make-the-asterisk-labeled-heatmap-for-f_vs_m-degs" class="nav-link" data-scroll-target="#to-make-the-asterisk-labeled-heatmap-for-f_vs_m-degs">to make the asterisk labeled heatmap for F_vs_M DEGs</a></li>
  </ul></li>
  <li><a href="#make-umaps-for-fvsm-degs" id="toc-make-umaps-for-fvsm-degs" class="nav-link" data-scroll-target="#make-umaps-for-fvsm-degs">make umaps for FvsM DEGs</a>
  <ul class="collapse">
  <li><a href="#make-heatmap-on-the-foldchanges-of-all-degs-between-f-vs.-m-in-iso-group" id="toc-make-heatmap-on-the-foldchanges-of-all-degs-between-f-vs.-m-in-iso-group" class="nav-link" data-scroll-target="#make-heatmap-on-the-foldchanges-of-all-degs-between-f-vs.-m-in-iso-group">make heatmap on the foldchanges of all DEGs between F vs.&nbsp;M in iso group</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">plot for the DEG number in each cell type</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div id="9f34d899-fceb-4bcb-bd17-f03fa38fa40a" class="cell" data-scrolled="true" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>library(Seurat)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>library(ggplot2)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>library(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>library(patchwork)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>library(stringr)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>library(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="de66ea70-8eee-4aaf-a992-1e60bc84c4dd" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="3">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">## https://drive.google.com/file/d/16Jq6w_KAWwI2rBN7SYBw-7bm5qwVoFyt/view?usp=drive_link</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>seu_obj <span class="op">&lt;-</span> readRDS(<span class="bu">file</span><span class="op">=</span><span class="st">"/project/zhuzhuzhang/lyang/scaleBio_RNA_rat_brain/out_mRatBN/merged_seurat_obj_celltype_annotated_2.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="d5646fea-9d27-4e44-9d1d-f8ac7a05c8ac" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="5">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">## https://drive.google.com/file/d/1sWhyhm5V5RlxMEJhZ1BUncDxaI2_tGXJ/view?usp=drive_link</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>pseudo_de_sex <span class="op">&lt;-</span> readRDS(<span class="bu">file</span> <span class="op">=</span> <span class="st">"./pseudo_sex_DESeq2_DEG_2.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="1c0148da-c577-407b-acf1-91ee854340ed" class="cell" data-execution_count="138">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">8</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">4</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="op">&lt;-</span> seu_obj<span class="op">@</span>meta.data <span class="op">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">#count(celltype) %&gt;%</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    mutate(group<span class="op">=</span>factor(group,levels<span class="op">=</span>c(<span class="st">"soc"</span>,<span class="st">"iso"</span>))) <span class="op">%&gt;%</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(celltype<span class="op">!=</span><span class="st">"NA"</span>) <span class="op">%&gt;%</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    ggplot(aes(x<span class="op">=</span>celltype))<span class="op">+</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    geom_bar(fill<span class="op">=</span><span class="st">"darkgrey"</span>)<span class="op">+</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    geom_text(stat<span class="op">=</span><span class="st">'count'</span>, aes(label<span class="op">=</span>..count..), vjust<span class="op">=-</span><span class="fl">0.2</span>,size<span class="op">=</span><span class="dv">2</span>)<span class="op">+</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    <span class="co">#geom_text(aes(label = n), vjust = 0.5,size=2)+</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    facet_grid(sex<span class="op">~</span>group)<span class="op">+</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    labs(x<span class="op">=</span><span class="st">""</span>,y<span class="op">=</span><span class="st">"# of cells"</span>)<span class="op">+</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>    theme_minimal()<span class="op">+</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    theme(axis.text.x<span class="op">=</span>element_blank(),</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>        <span class="co">#axis.text.x=element_text(angle=90,hjust=1),</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>         plot.margin <span class="op">=</span> unit(c(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>), <span class="st">"cm"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="e5f1520b-3d28-4e0c-8491-32a4ed864d6e" class="cell" data-execution_count="131">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">8</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">2</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="op">&lt;-</span> sapply(pseudo_de_sex,nrow) <span class="op">%&gt;%</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="im">as</span>.data.frame() <span class="op">%&gt;%</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>setNames(<span class="st">"gene_num"</span>) <span class="op">%&gt;%</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>mutate(name<span class="op">=</span>row.names(.)) <span class="op">%&gt;%</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>mutate(celltype<span class="op">=</span>str_split(name,<span class="st">"_"</span>,simplify <span class="op">=</span> T)[,<span class="dv">1</span>],</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>      group<span class="op">=</span>str_split(name,<span class="st">"_"</span>,simplify <span class="op">=</span> T)[,<span class="dv">2</span>]) <span class="op">%&gt;%</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>mutate(celltype<span class="op">=</span>factor(celltype,levels<span class="op">=</span>levels(seu_obj<span class="op">@</span>meta.data$celltype)[<span class="dv">1</span>:<span class="dv">15</span>])) <span class="op">%&gt;%</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>mutate(group<span class="op">=</span>factor(group,levels<span class="op">=</span>c(<span class="st">"soc"</span>,<span class="st">"iso"</span>))) <span class="op">%&gt;%</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>ggplot(aes(x<span class="op">=</span>celltype,y<span class="op">=</span>gene_num))<span class="op">+</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>    geom_bar(stat <span class="op">=</span> <span class="st">"identity"</span>,fill<span class="op">=</span><span class="st">"darkgrey"</span>)<span class="op">+</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>    geom_text(aes(label <span class="op">=</span> gene_num),</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>              position <span class="op">=</span> position_stack(vjust <span class="op">=</span> <span class="fl">1.02</span>),size<span class="op">=</span><span class="dv">2</span>)<span class="op">+</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>    facet_grid(.<span class="op">~</span>group)<span class="op">+</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>    theme_minimal()<span class="op">+</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>    theme(axis.text.x <span class="op">=</span> element_text(size<span class="op">=</span><span class="dv">8</span>,angle<span class="op">=</span><span class="dv">30</span>,hjust<span class="op">=</span><span class="dv">1</span>,vjust<span class="op">=</span><span class="dv">1</span>))<span class="op">+</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>    labs(y<span class="op">=</span><span class="st">"gene number"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="d633b930-4ee8-40b9-89be-9aa54639f1ef" class="cell" data-execution_count="140">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>options(<span class="bu">repr</span>.plot.width <span class="op">=</span> <span class="dv">8</span>, <span class="bu">repr</span>.plot.height <span class="op">=</span> <span class="dv">5</span>, <span class="bu">repr</span>.plot.res <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>p1<span class="op">/</span>p2 <span class="op">+</span> plot_layout(heights <span class="op">=</span> c(<span class="dv">2</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="sex_DEG_plot_files/figure-html/cell-7-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<section id="violin-plots-for-genes" class="level2">
<h2 class="anchored" data-anchor-id="violin-plots-for-genes">violin plots for genes</h2>
<div id="5c0be204-255d-4ea3-9ac9-e569d221d2ce" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="33">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>make_vlnplot <span class="op">&lt;-</span> </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>function(seuObject,genes<span class="op">=</span>NULL,celltype<span class="op">=</span>NULL,label_df<span class="op">=</span>NULL){</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    VlnPlot(seuObject, </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>        features <span class="op">=</span> genes,</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>        idents<span class="op">=</span>celltype,</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>        split.by <span class="op">=</span> <span class="st">"sex"</span>,</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>        pt.size <span class="op">=</span> <span class="fl">0.0</span>,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>        <span class="co">#split.plot = T,</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>        log<span class="op">=</span>FALSE  )<span class="op">&amp;</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>        scale_fill_manual(values<span class="op">=</span>c(<span class="st">"M"</span><span class="op">=</span><span class="st">"#8ecae6"</span>,<span class="st">"F"</span><span class="op">=</span><span class="st">"#ffe5ec"</span>)) <span class="op">&amp;</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>        annotate(<span class="st">"text"</span>,x<span class="op">=</span>label_df$celltype,y<span class="op">=</span>Inf,label<span class="op">=</span>label_df$P_label,vjust<span class="op">=</span><span class="dv">1</span>,hjust<span class="op">=</span><span class="fl">0.5</span>) <span class="op">&amp;</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>        <span class="co">#stat_summary(fun = median, fun.min = median, fun.max = median,</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>        <span class="co">#         geom = "crossbar", </span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>        <span class="co">#         width = 0.6,</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>        <span class="co">#         position = position_dodge(width = .70)) &amp;</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>                xlab(<span class="st">""</span>) <span class="op">&amp;</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>                ylab(<span class="st">"Gene Expression"</span>) <span class="op">&amp;</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>          stat_summary(fun <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>               geom <span class="op">=</span> <span class="st">"point"</span>,</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>               color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>               position <span class="op">=</span> position_dodge(width <span class="op">=</span> <span class="fl">.90</span>)) <span class="op">&amp;</span></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>    labs(caption <span class="op">=</span> paste(<span class="st">"adj.p: "</span>,<span class="st">"**** &lt;0.0001"</span>,<span class="st">"*** &lt;0.001"</span>,<span class="st">"** &lt;0.01"</span>,<span class="st">"* &lt;0.05"</span>)) <span class="op">&amp;</span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>    guides(fill <span class="op">=</span> guide_legend(override.aes <span class="op">=</span> <span class="bu">list</span>(linetype <span class="op">=</span> <span class="dv">0</span>,size <span class="op">=</span> <span class="dv">0</span>)),</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>               color <span class="op">=</span> guide_legend(override.aes <span class="op">=</span> <span class="bu">list</span>(linetype <span class="op">=</span> <span class="dv">0</span>,size<span class="op">=</span><span class="dv">0</span>)))</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="92cfaf2c-6794-4220-b230-3e08cdc65dbf" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>Idents(seu_obj) <span class="op">&lt;-</span> <span class="st">"celltype"</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>seu_obj<span class="op">@</span>meta.data$group <span class="op">&lt;-</span> factor(seu_obj<span class="op">@</span>meta.data$group,levels<span class="op">=</span>c(<span class="st">"soc"</span>,<span class="st">"iso"</span>))</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>seu_obj<span class="op">@</span>meta.data$sex <span class="op">&lt;-</span> factor(seu_obj<span class="op">@</span>meta.data$sex,levels<span class="op">=</span>c(<span class="st">"F"</span>,<span class="st">"M"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fbc94054-4a54-47b8-849f-2e518b3d20ee" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="7">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>FvsM_DEG_ls <span class="op">&lt;-</span> read.table(<span class="st">"FvsM_DEG_total_list.xls"</span>,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,stringsAsFactors <span class="op">=</span> F,header<span class="op">=</span>T)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>FvsM_DEG_ls$P_label<span class="op">&lt;-</span> </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    symnum(FvsM_DEG_ls$p_val_adj, cutpoints <span class="op">=</span> c(<span class="dv">0</span>, <span class="fl">0.0001</span>, <span class="fl">0.001</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span>,<span class="dv">1</span>),</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>       symbols <span class="op">=</span> c(<span class="st">"****"</span>, <span class="st">"***"</span>, <span class="st">"**"</span>, <span class="st">"*"</span>, <span class="st">"ns"</span>),</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>       abbr.colnames <span class="op">=</span> FALSE, na <span class="op">=</span> <span class="st">""</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>FvsM_DEG_ls$celltype <span class="op">&lt;-</span> str_split(FvsM_DEG_ls$new_clumn,<span class="st">"_"</span>,simplify<span class="op">=</span>T)[,<span class="dv">1</span>]</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>FvsM_DEG_ls$group <span class="op">&lt;-</span> str_split(FvsM_DEG_ls$new_clumn,<span class="st">"_"</span>,simplify<span class="op">=</span>T)[,<span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="e113567a-e9f4-403f-8126-5967a8740529" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>table(FvsM_DEG_ls$group)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>
iso soc 
 82 211 </code></pre>
</div>
</div>
<div id="751763c8-932e-42ef-814a-d317ee44d510" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="9">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>soc_gene_for_plot <span class="op">&lt;-</span> </span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>FvsM_DEG_ls <span class="op">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    select(gene,new_clumn) <span class="op">%&gt;%</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    mutate(celltype<span class="op">=</span>str_split(new_clumn,<span class="st">"_"</span>,simplify<span class="op">=</span>T)[,<span class="dv">1</span>],</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>          group<span class="op">=</span>str_split(new_clumn,<span class="st">"_"</span>,simplify<span class="op">=</span>T)[,<span class="dv">2</span>]) <span class="op">%&gt;%</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(group<span class="op">==</span><span class="st">"soc"</span>)<span class="op">%&gt;%</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    group_by(gene) <span class="op">%&gt;%</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>    summarise(celltype <span class="op">=</span> str_c(celltype, collapse<span class="op">=</span><span class="st">","</span>))</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>iso_gene_for_plot <span class="op">&lt;-</span> </span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>FvsM_DEG_ls <span class="op">%&gt;%</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>    select(gene,new_clumn) <span class="op">%&gt;%</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>    mutate(celltype<span class="op">=</span>str_split(new_clumn,<span class="st">"_"</span>,simplify<span class="op">=</span>T)[,<span class="dv">1</span>],</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>          group<span class="op">=</span>str_split(new_clumn,<span class="st">"_"</span>,simplify<span class="op">=</span>T)[,<span class="dv">2</span>]) <span class="op">%&gt;%</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(group<span class="op">==</span><span class="st">"iso"</span>)<span class="op">%&gt;%</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>    group_by(gene) <span class="op">%&gt;%</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>    summarise(celltype <span class="op">=</span> str_c(celltype, collapse<span class="op">=</span><span class="st">","</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="6ca07ef3" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="10">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>gene_ls <span class="op">&lt;-</span> c(<span class="st">"Ndst4"</span>,<span class="st">"Cdh12"</span>,<span class="st">"Gpc5"</span>,<span class="st">"Gpc6"</span>,<span class="st">"Lingo2"</span>,<span class="st">"Nrxn3"</span>,<span class="st">"Ralyl"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="48071d30-4562-4f33-86e9-ca5ccccdb742" class="cell" data-scrolled="true">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>library(ggpubr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="891be948-9b95-45b5-8a3b-21b9bb067d11" class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>texto <span class="op">&lt;-</span> textGrob(</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  label <span class="op">=</span> c(<span class="st">"adj.p"</span>,<span class="st">"**** &lt;0.0001"</span>,<span class="st">"*** &lt;0.001"</span>,<span class="st">"** &lt;0.01"</span>,<span class="st">"* &lt;0.05"</span>),</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  x <span class="op">=</span> c(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>),</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  y <span class="op">=</span> c(<span class="dv">0</span>,<span class="fl">0.2</span>,<span class="fl">0.4</span>,<span class="fl">0.6</span>,<span class="fl">0.8</span>),</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  gp <span class="op">=</span> gpar(lwd <span class="op">=</span> <span class="dv">2</span>, fontsize <span class="op">=</span> <span class="dv">5</span>),</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="d89e0fa9" class="cell" data-scrolled="true" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>iso_FvsM_DEGs <span class="op">&lt;-</span> unique(iso_gene_for_plot$gene)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>pdf(<span class="bu">file</span> <span class="op">=</span> <span class="st">"./Vlnplot_for_iso_FvsM_DEGs.pdf"</span>,width<span class="op">=</span><span class="dv">8</span>,height<span class="op">=</span><span class="dv">4</span>,onefile<span class="op">=</span>TRUE)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>seuObject <span class="op">&lt;-</span> subset(seu_obj,subset<span class="op">=</span>group<span class="op">==</span><span class="st">"iso"</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>seuObject<span class="op">@</span>meta.data$sex <span class="op">&lt;-</span> factor(seuObject<span class="op">@</span>meta.data$sex,levels<span class="op">=</span>c(<span class="st">"M"</span>,<span class="st">"F"</span>))</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="kw">in</span> seq_len(length(iso_FvsM_DEGs))){</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(paste0(<span class="st">"prcocessing gene: "</span>,iso_FvsM_DEGs[i]))</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    label_df <span class="op">&lt;-</span> </span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>        FvsM_DEG_ls <span class="op">%&gt;%</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>        <span class="bu">filter</span>(gene<span class="op">==</span>iso_FvsM_DEGs[i] <span class="op">&amp;</span> group<span class="op">==</span><span class="st">"iso"</span>)</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(nrow(label_df))</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(nrow(label_df)<span class="op">&gt;</span><span class="dv">0</span>){</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>        p <span class="op">&lt;-</span> make_vlnplot(seuObject,</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>                      genes<span class="op">=</span>iso_FvsM_DEGs[i],</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>                      celltype<span class="op">=</span>levels(seu_obj<span class="op">@</span>meta.data$celltype)[<span class="dv">1</span>:<span class="dv">14</span>],</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>                      label_df<span class="op">=</span>label_df</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>                 )</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(p)</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>dev.off()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="save-the-data-matrix-for-all-degs-between-f-and-m" class="level2">
<h2 class="anchored" data-anchor-id="save-the-data-matrix-for-all-degs-between-f-and-m">save the data matrix for all DEGs between F and M</h2>
<div id="fc532411-9648-4b4b-a7cb-12fc3f0dd179" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>iso_compare <span class="op">&lt;-</span> pseudo_de_sex[names(pseudo_de_sex)[grep(<span class="st">"iso"</span>,names(pseudo_de_sex))]]</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>soc_compare <span class="op">&lt;-</span> pseudo_de_sex[names(pseudo_de_sex)[grep(<span class="st">"soc"</span>,names(pseudo_de_sex))]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="get-the-deg-list-for-iso-group-comparison" class="level4">
<h4 class="anchored" data-anchor-id="get-the-deg-list-for-iso-group-comparison">get the DEG list for iso group comparison</h4>
<div id="f8db934e-5ba6-4f48-ad4a-ac43dbbff871" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>DEG_ls<span class="op">&lt;-</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>lapply(seq_len(length(names(iso_compare))), function(x) {</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="bu">sum</span>(iso_compare[[x]]$p_val_adj <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="bu">abs</span>(iso_compare[[x]]$avg_log2FC) <span class="op">&gt;</span> <span class="fl">0.3</span>,na.rm <span class="op">=</span> TRUE) <span class="op">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>        iso_compare[[x]] <span class="op">%&gt;%</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>            <span class="bu">filter</span>(p_val_adj <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="bu">abs</span>(avg_log2FC) <span class="op">&gt;</span> <span class="fl">0.3</span>)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>names(DEG_ls) <span class="op">&lt;-</span> str_split(names(iso_compare),<span class="st">"_"</span>,simplify <span class="op">=</span> T)[,<span class="dv">1</span>]</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>DEG_ls <span class="op">&lt;-</span> DEG_ls[<span class="op">!</span>sapply(DEG_ls,<span class="kw">is</span>.null)]</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="co">## load gene coordinates information</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>gene_coord <span class="op">&lt;-</span> read.table(<span class="st">"/project/zhuzhuzhang/lyang/references/mRatBN7.2_NCBI/gene_coordinates"</span>,header<span class="op">=</span>F,stringsAsFactors <span class="op">=</span> F) <span class="op">%&gt;%</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>setNames(c(<span class="st">"chr"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"strand"</span>,<span class="st">"gene"</span>))</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>do.call(rbind,Map(cbind, DEG_ls, new_clumn <span class="op">=</span> names(DEG_ls))) <span class="op">%&gt;%</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>  mutate(gene <span class="op">=</span> str_split(row.names(.), <span class="st">"</span><span class="ch">\\</span><span class="st">."</span>, simplify <span class="op">=</span> TRUE)[, <span class="dv">2</span>]) <span class="op">%&gt;%</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(gene <span class="op">!=</span> <span class="st">""</span>) <span class="op">%&gt;%</span></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>    left_join(gene_coord,by<span class="op">=</span>c(<span class="st">"gene"</span><span class="op">=</span><span class="st">"gene"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>    write.table(<span class="st">"iso_FvsM_DEG_total_list.txt"</span>,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,quote<span class="op">=</span>F,row.names<span class="op">=</span>F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="get-the-deg-list-for-soc-group-comparison" class="level4">
<h4 class="anchored" data-anchor-id="get-the-deg-list-for-soc-group-comparison">get the DEG list for soc group comparison</h4>
<div id="19cfeb56-9c28-43b0-bc70-4998ca88ea67" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>DEG_ls<span class="op">&lt;-</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>lapply(seq_len(length(names(soc_compare))), function(x) {</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="bu">sum</span>(soc_compare[[x]]$p_val_adj <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="bu">abs</span>(soc_compare[[x]]$avg_log2FC) <span class="op">&gt;</span> <span class="fl">0.3</span>,na.rm <span class="op">=</span> TRUE) <span class="op">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>        soc_compare[[x]] <span class="op">%&gt;%</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>            <span class="bu">filter</span>(p_val_adj <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="bu">abs</span>(avg_log2FC) <span class="op">&gt;</span> <span class="fl">0.3</span>)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>names(DEG_ls) <span class="op">&lt;-</span> str_split(names(soc_compare),<span class="st">"_"</span>,simplify <span class="op">=</span> T)[,<span class="dv">1</span>]</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>DEG_ls <span class="op">&lt;-</span> DEG_ls[<span class="op">!</span>sapply(DEG_ls,<span class="kw">is</span>.null)]</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>do.call(rbind,Map(cbind, DEG_ls, new_clumn <span class="op">=</span> names(DEG_ls))) <span class="op">%&gt;%</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>  mutate(gene <span class="op">=</span> str_split(row.names(.), <span class="st">"</span><span class="ch">\\</span><span class="st">."</span>, simplify <span class="op">=</span> TRUE)[, <span class="dv">2</span>]) <span class="op">%&gt;%</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(gene <span class="op">!=</span> <span class="st">""</span>) <span class="op">%&gt;%</span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>    left_join(gene_coord,by<span class="op">=</span>c(<span class="st">"gene"</span><span class="op">=</span><span class="st">"gene"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>    write.table(<span class="st">"soc_FvsM_DEG_total_list.txt"</span>,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,quote<span class="op">=</span>F,row.names<span class="op">=</span>F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="5ff8dd99-9c2f-44ae-b50b-c1ae3ab8f156" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>soc_FvsM_DEG <span class="op">&lt;-</span> read.table(<span class="st">"soc_FvsM_DEG_total_list.txt"</span>,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,header<span class="op">=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="c6265f08-7799-4232-aa2f-fc6ccabecbc2" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>iso_FvsM_DEG <span class="op">&lt;-</span> read.table(<span class="st">"iso_FvsM_DEG_total_list.txt"</span>,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,header<span class="op">=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="save-the-fc-data-for-all-celltypes-for-all-degs" class="level2">
<h2 class="anchored" data-anchor-id="save-the-fc-data-for-all-celltypes-for-all-degs">save the FC data for all celltypes for all DEGs</h2>
<section id="to-make-the-asterisk-labeled-heatmap-for-f_vs_m-degs" class="level3">
<h3 class="anchored" data-anchor-id="to-make-the-asterisk-labeled-heatmap-for-f_vs_m-degs">to make the asterisk labeled heatmap for F_vs_M DEGs</h3>
<div id="e187b50b-e76a-4c1a-8b69-6e02fb4e3c81" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">## gather fold change information together</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>DEG_ls<span class="op">&lt;-</span> </span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>lapply(seq_len(length(names(iso_compare))), function(x) {</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>        iso_compare[[x]] <span class="op">%&gt;%</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>            <span class="bu">filter</span>(row.names(.) <span class="op">%</span><span class="kw">in</span><span class="op">%</span> unique(iso_FvsM_DEG$gene))</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>names(DEG_ls) <span class="op">&lt;-</span> str_split(names(iso_compare),<span class="st">"_"</span>,simplify <span class="op">=</span> T)[,<span class="dv">1</span>]</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>all_DEG_gene <span class="op">&lt;-</span> do.call(rbind,Map(cbind, DEG_ls, new_clumn <span class="op">=</span> names(DEG_ls))) <span class="op">%&gt;%</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>  mutate(gene <span class="op">=</span> str_split(row.names(.), <span class="st">"</span><span class="ch">\\</span><span class="st">."</span>, simplify <span class="op">=</span> TRUE)[, <span class="dv">2</span>]) <span class="op">%&gt;%</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(gene <span class="op">!=</span> <span class="st">""</span>)</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>all_DEG_gene <span class="op">%&gt;%</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>    select(new_clumn,avg_log2FC,gene) <span class="op">%&gt;%</span></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>    tidyr::pivot_wider(names_from <span class="op">=</span> new_clumn, values_from <span class="op">=</span> avg_log2FC) <span class="op">%&gt;%</span></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>    tibble::column_to_rownames(var<span class="op">=</span><span class="st">"gene"</span>) <span class="op">%&gt;%</span></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>    write.table(<span class="bu">file</span><span class="op">=</span><span class="st">"./iso_FvsM_DEG_FC_celltype_matrix.tsv"</span>,quote<span class="op">=</span>F,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,row.names<span class="op">=</span>T)</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>all_DEG_gene <span class="op">%&gt;%</span></span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>    select(new_clumn,p_val_adj,gene) <span class="op">%&gt;%</span></span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>    tidyr::pivot_wider(names_from <span class="op">=</span> new_clumn, values_from <span class="op">=</span> p_val_adj) <span class="op">%&gt;%</span></span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>    tibble::column_to_rownames(var<span class="op">=</span><span class="st">"gene"</span>) <span class="op">%&gt;%</span></span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>    write.table(<span class="bu">file</span><span class="op">=</span><span class="st">"./iso_FvsM_DEG_adjP_celltype_matrix.tsv"</span>,quote<span class="op">=</span>F,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,row.names<span class="op">=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>same for social grouped comparison</p>
<div id="58cb5d07-bbca-4d43-ba17-8e14b62a55a4" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">## gather fold change information together</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>DEG_ls<span class="op">&lt;-</span> </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>lapply(seq_len(length(names(soc_compare))), function(x) {</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>        soc_compare[[x]] <span class="op">%&gt;%</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>            <span class="bu">filter</span>(row.names(.) <span class="op">%</span><span class="kw">in</span><span class="op">%</span> unique(soc_FvsM_DEG$gene))</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>names(DEG_ls) <span class="op">&lt;-</span> str_split(names(soc_compare),<span class="st">"_"</span>,simplify <span class="op">=</span> T)[,<span class="dv">1</span>]</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>all_DEG_gene <span class="op">&lt;-</span> do.call(rbind,Map(cbind, DEG_ls, new_clumn <span class="op">=</span> names(DEG_ls))) <span class="op">%&gt;%</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  mutate(gene <span class="op">=</span> str_split(row.names(.), <span class="st">"</span><span class="ch">\\</span><span class="st">."</span>, simplify <span class="op">=</span> TRUE)[, <span class="dv">2</span>]) <span class="op">%&gt;%</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(gene <span class="op">!=</span> <span class="st">""</span>)</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>all_DEG_gene <span class="op">%&gt;%</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>    select(new_clumn,avg_log2FC,gene) <span class="op">%&gt;%</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>    tidyr::pivot_wider(names_from <span class="op">=</span> new_clumn, values_from <span class="op">=</span> avg_log2FC) <span class="op">%&gt;%</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>    tibble::column_to_rownames(var<span class="op">=</span><span class="st">"gene"</span>) <span class="op">%&gt;%</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>    write.table(<span class="bu">file</span><span class="op">=</span><span class="st">"./soc_FvsM_DEG_FC_celltype_matrix.tsv"</span>,quote<span class="op">=</span>F,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,row.names<span class="op">=</span>T)</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>all_DEG_gene <span class="op">%&gt;%</span></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>    select(new_clumn,p_val_adj,gene) <span class="op">%&gt;%</span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>    tidyr::pivot_wider(names_from <span class="op">=</span> new_clumn, values_from <span class="op">=</span> p_val_adj) <span class="op">%&gt;%</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>    tibble::column_to_rownames(var<span class="op">=</span><span class="st">"gene"</span>) <span class="op">%&gt;%</span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>    write.table(<span class="bu">file</span><span class="op">=</span><span class="st">"./soc_FvsM_DEG_adjP_celltype_matrix.tsv"</span>,quote<span class="op">=</span>F,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,row.names<span class="op">=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="make-umaps-for-fvsm-degs" class="level2">
<h2 class="anchored" data-anchor-id="make-umaps-for-fvsm-degs">make umaps for FvsM DEGs</h2>
<div id="3855539b-95e7-4586-bd94-a00736370fcb" class="cell" data-execution_count="99">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>iso_FvsM_DEG <span class="op">&lt;-</span> </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>    FvsM_DEG_ls <span class="op">%&gt;%</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(group<span class="op">==</span><span class="st">"iso"</span>) <span class="op">%&gt;%</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>    pull(gene) <span class="op">%&gt;%</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    unique()</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>pdf(<span class="bu">file</span> <span class="op">=</span> <span class="st">"./iso_FvsM_DEG_umap.pdf"</span>,width<span class="op">=</span><span class="dv">10</span>,height<span class="op">=</span><span class="dv">5</span>,onefile<span class="op">=</span>TRUE)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="kw">in</span> seq_len(length(iso_FvsM_DEG))){</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>    p <span class="op">&lt;-</span> FeaturePlot(seu_obj, features <span class="op">=</span> iso_FvsM_DEG[i],split.by<span class="op">=</span><span class="st">"sex"</span>) <span class="op">+</span> </span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>            theme(legend.position <span class="op">=</span> c(<span class="op">-</span><span class="fl">0.35</span>,<span class="fl">0.2</span>))<span class="op">+</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>            theme(legend.text<span class="op">=</span>element_text(size<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>))</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(p)</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>dev.off()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<strong>png:</strong> 2
</div>
</div>
<div id="780a3804-6613-4d02-acf1-48ec5071f2c6" class="cell" data-execution_count="108">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>soc_FvsM_DEG <span class="op">&lt;-</span> </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>    FvsM_DEG_ls <span class="op">%&gt;%</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(group<span class="op">==</span><span class="st">"soc"</span>) <span class="op">%&gt;%</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    pull(gene) <span class="op">%&gt;%</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    unique()</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>soc_FvsM_DEG <span class="op">&lt;-</span> </span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>    FvsM_DEG_ls <span class="op">%&gt;%</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(group<span class="op">==</span><span class="st">"soc"</span>) <span class="op">%&gt;%</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>    pull(gene) <span class="op">%&gt;%</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>    unique()</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>pdf(<span class="bu">file</span> <span class="op">=</span> <span class="st">"./soc_FvsM_DEG_umap.pdf"</span>,width<span class="op">=</span><span class="dv">10</span>,height<span class="op">=</span><span class="dv">5</span>,onefile<span class="op">=</span>TRUE)</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="kw">in</span> seq_len(length(soc_FvsM_DEG))){</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>    p <span class="op">&lt;-</span> FeaturePlot(seu_obj, features <span class="op">=</span> soc_FvsM_DEG[i],split.by<span class="op">=</span><span class="st">"sex"</span>) <span class="op">+</span> </span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>            theme(legend.position <span class="op">=</span> c(<span class="op">-</span><span class="fl">0.35</span>,<span class="fl">0.2</span>))<span class="op">+</span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>            theme(legend.text<span class="op">=</span>element_text(size<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>))</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(p)</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>dev.off()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<strong>png:</strong> 2
</div>
</div>
<section id="make-heatmap-on-the-foldchanges-of-all-degs-between-f-vs.-m-in-iso-group" class="level3">
<h3 class="anchored" data-anchor-id="make-heatmap-on-the-foldchanges-of-all-degs-between-f-vs.-m-in-iso-group">make heatmap on the foldchanges of all DEGs between F vs.&nbsp;M in iso group</h3>
<div id="78ae373a-e207-4030-a9e5-a12fffafa924" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>iso_FvsM_DEG_FC <span class="op">&lt;-</span> read.table(<span class="st">"./iso_FvsM_DEG_FC_celltype_matrix.tsv"</span>, header<span class="op">=</span>T,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,row.names<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>iso_FvsM_DEG_P <span class="op">&lt;-</span> read.table(<span class="st">"./iso_FvsM_DEG_adjP_celltype_matrix.tsv"</span>, header<span class="op">=</span>T,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,row.names<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>set cutoffs for asterisk label of significance</p>
<div id="ad6ce736-cf6f-4479-8649-14e23bc16405" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>rm.neg<span class="op">&lt;-</span>plyr::colwise(function(x){</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span>(ifelse(x <span class="op">&lt;</span> <span class="fl">0.0001</span>,<span class="st">"****"</span>, </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>                ifelse(x<span class="op">&lt;</span><span class="fl">0.001</span>,<span class="st">"***"</span>,</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>                      ifelse(x<span class="op">&lt;</span><span class="fl">0.01</span>,<span class="st">"**"</span>,</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>                            ifelse(x<span class="op">&lt;</span><span class="fl">0.05</span>,<span class="st">"*"</span>,<span class="st">""</span>)))))})</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>p_sign_df <span class="op">&lt;-</span> rm.neg(iso_FvsM_DEG_P)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="80c9302a-38d6-4b88-b6e9-b83ebdc6c25a" class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co">## make sure the column names are consistent in the p-value and foldchange table</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>row.names(p_sign_df) <span class="op">&lt;-</span> row.names(iso_FvsM_DEG_P)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>colnames(p_sign_df) <span class="op">&lt;-</span> c(<span class="st">"L2/3 IT"</span>, <span class="st">"L4 IT"</span>, <span class="st">"L5 IT"</span>, <span class="st">"L6 IT"</span>, <span class="st">"L5/6 NP"</span>, <span class="st">"L6 CT"</span>, <span class="st">"L5 ET"</span>, <span class="st">"CGE-Inh"</span>, <span class="st">"Sst-Inh"</span>, <span class="st">"Pvalb-Inh"</span>, <span class="st">"ASC"</span>, <span class="st">"ODC"</span>, <span class="st">"OPC"</span>, <span class="st">"Endo"</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>colnames(iso_FvsM_DEG_FC) <span class="op">&lt;-</span> c(<span class="st">"L2/3 IT"</span>, <span class="st">"L4 IT"</span>, <span class="st">"L5 IT"</span>, <span class="st">"L6 IT"</span>, <span class="st">"L5/6 NP"</span>, <span class="st">"L6 CT"</span>, <span class="st">"L5 ET"</span>, <span class="st">"CGE-Inh"</span>, <span class="st">"Sst-Inh"</span>, <span class="st">"Pvalb-Inh"</span>, <span class="st">"ASC"</span>, <span class="st">"ODC"</span>, <span class="st">"OPC"</span>, <span class="st">"Endo"</span>)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>p_sign_df[<span class="kw">is</span>.na(p_sign_df)] <span class="op">&lt;-</span> <span class="st">""</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>p_sign_df[<span class="bu">abs</span>(iso_FvsM_DEG_FC)<span class="op">&lt;</span><span class="fl">0.3</span>] <span class="op">&lt;-</span> <span class="st">""</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="5fdc0ce0-e8ef-45d3-b547-264e85636b35" class="cell" data-scrolled="true">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>library(circlize)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>library(ComplexHeatmap)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="f226200a-d6d3-4061-8226-06e89eb1edba" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>col_fun <span class="op">=</span> colorRamp2(c(<span class="op">-</span><span class="dv">5</span>, <span class="dv">0</span>, <span class="dv">5</span>), c(<span class="st">"#0077b6"</span>, <span class="st">"white"</span>, <span class="st">"#e63946"</span>))</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> Heatmap(iso_FvsM_DEG_FC, name <span class="op">=</span> <span class="st">"iso F/M</span><span class="ch">\n</span><span class="st">avg_logFC"</span>, col <span class="op">=</span> col_fun,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>        cluster_columns <span class="op">=</span> F,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>        show_row_dend <span class="op">=</span> F,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>        heatmap_legend_param <span class="op">=</span> <span class="bu">list</span>(</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>            legend_direction <span class="op">=</span> <span class="st">"horizontal"</span>, </span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>        legend_width <span class="op">=</span> unit(<span class="dv">6</span>, <span class="st">"cm"</span>)),</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>        cell_fun <span class="op">=</span> function(j, i, x, y, width, height, fill) {</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>            grid.text(sprintf(<span class="st">"</span><span class="sc">%s</span><span class="st">"</span>, p_sign_df[i, j]), x, y, gp <span class="op">=</span> gpar(fontsize <span class="op">=</span> <span class="dv">10</span>))</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>lgd1 <span class="op">=</span> Legend(labels<span class="op">=</span>c(<span class="st">"&lt;0.0001"</span>,<span class="st">"&lt;0.001"</span>,<span class="st">"&lt;0.01"</span>,<span class="st">"&lt;0.05"</span>),pch <span class="op">=</span> c(<span class="st">"**** "</span>,<span class="st">"*** "</span>,<span class="st">"** "</span>,<span class="st">"* "</span>), title <span class="op">=</span> <span class="st">"adj.p"</span>, <span class="bu">type</span> <span class="op">=</span> <span class="st">"points"</span>, </span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>    background <span class="op">=</span> <span class="st">"white"</span>,direction <span class="op">=</span> <span class="st">"horizontal"</span>)</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>lgd2 <span class="op">=</span> Legend(labels<span class="op">=</span><span class="st">"gene not detected"</span>,legend_gp <span class="op">=</span> gpar(fill <span class="op">=</span> <span class="st">"grey"</span>), title <span class="op">=</span> <span class="st">"NA"</span>)</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a>pdf(<span class="st">"iso_FvsM_DEG_logFC_heatmap_2.pdf"</span>,width<span class="op">=</span><span class="dv">7</span>,height<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>draw(p,annotation_legend_list <span class="op">=</span> <span class="bu">list</span>(lgd1,lgd2),heatmap_legend_side<span class="op">=</span><span class="st">"bottom"</span>, annotation_legend_side<span class="op">=</span><span class="st">"right"</span>,</span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>           legend_grouping <span class="op">=</span> <span class="st">"original"</span>)</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>dev.off()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“The input is a data frame-like object, convert it to a matrix.”</code></pre>
</div>
<div class="cell-output cell-output-display">
<strong>png:</strong> 2
</div>
</div>
<section id="same-plot-for-soc-group" class="level4">
<h4 class="anchored" data-anchor-id="same-plot-for-soc-group">Same plot for soc group</h4>
<div id="8d34a7c9-2209-4ad6-a5fc-8e6b464a4a2b" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>soc_FvsM_DEG_FC <span class="op">&lt;-</span> read.table(<span class="st">"./soc_FvsM_DEG_FC_celltype_matrix.tsv"</span>, header<span class="op">=</span>T,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,row.names<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>soc_FvsM_DEG_P <span class="op">&lt;-</span> read.table(<span class="st">"./soc_FvsM_DEG_adjP_celltype_matrix.tsv"</span>, header<span class="op">=</span>T,sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,row.names<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>p_sign_df <span class="op">&lt;-</span> rm.neg(soc_FvsM_DEG_P)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>row.names(p_sign_df) <span class="op">&lt;-</span> row.names(soc_FvsM_DEG_P)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>colnames(p_sign_df) <span class="op">&lt;-</span> c(<span class="st">"L2/3 IT"</span>, <span class="st">"L4 IT"</span>, <span class="st">"L5 IT"</span>, <span class="st">"L6 IT"</span>, <span class="st">"L5/6 NP"</span>, <span class="st">"L6 CT"</span>, <span class="st">"L5 ET"</span>, <span class="st">"CGE-Inh"</span>, <span class="st">"Sst-Inh"</span>, <span class="st">"Pvalb-Inh"</span>, <span class="st">"ASC"</span>, <span class="st">"ODC"</span>, <span class="st">"OPC"</span>, <span class="st">"Endo"</span>)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>colnames(soc_FvsM_DEG_FC) <span class="op">&lt;-</span> c(<span class="st">"L2/3 IT"</span>, <span class="st">"L4 IT"</span>, <span class="st">"L5 IT"</span>, <span class="st">"L6 IT"</span>, <span class="st">"L5/6 NP"</span>, <span class="st">"L6 CT"</span>, <span class="st">"L5 ET"</span>, <span class="st">"CGE-Inh"</span>, <span class="st">"Sst-Inh"</span>, <span class="st">"Pvalb-Inh"</span>, <span class="st">"ASC"</span>, <span class="st">"ODC"</span>, <span class="st">"OPC"</span>, <span class="st">"Endo"</span>)</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>p_sign_df[<span class="kw">is</span>.na(p_sign_df)] <span class="op">&lt;-</span> <span class="st">""</span></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>p_sign_df[<span class="bu">abs</span>(soc_FvsM_DEG_FC)<span class="op">&lt;</span><span class="fl">0.3</span>] <span class="op">&lt;-</span> <span class="st">""</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="c57db2a4-8ed3-4fb7-9618-10d497310107" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>col_fun <span class="op">=</span> colorRamp2(c(<span class="op">-</span><span class="dv">5</span>, <span class="dv">0</span>, <span class="dv">5</span>), c(<span class="st">"#0077b6"</span>, <span class="st">"white"</span>, <span class="st">"#e63946"</span>))</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> Heatmap(soc_FvsM_DEG_FC, name <span class="op">=</span> <span class="st">"soc F/M</span><span class="ch">\n</span><span class="st">avg_logFC"</span>, col <span class="op">=</span> col_fun,</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>        cluster_columns <span class="op">=</span> F,</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>        show_row_dend <span class="op">=</span> F,</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>        heatmap_legend_param <span class="op">=</span> <span class="bu">list</span>(</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>            legend_direction <span class="op">=</span> <span class="st">"horizontal"</span>, </span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>        legend_width <span class="op">=</span> unit(<span class="dv">6</span>, <span class="st">"cm"</span>)),</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>        row_names_gp <span class="op">=</span> grid::gpar(fontsize <span class="op">=</span> <span class="dv">4</span>),</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>        cell_fun <span class="op">=</span> function(j, i, x, y, width, height, fill) {</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>            gb <span class="op">=</span> textGrob(<span class="st">"*"</span>)</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>            gb_w <span class="op">=</span> convertWidth(grobWidth(gb), <span class="st">"mm"</span>)</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>            gb_h <span class="op">=</span> convertHeight(grobHeight(gb), <span class="st">"mm"</span>)</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>            grid.text(sprintf(<span class="st">"</span><span class="sc">%s</span><span class="st">"</span>, p_sign_df[i, j]), x, y <span class="op">-</span> gb_h<span class="op">*</span><span class="fl">0.5</span> <span class="op">+</span> gb_w<span class="op">*</span><span class="fl">0.4</span>, gp <span class="op">=</span> gpar(fontsize <span class="op">=</span> <span class="dv">10</span>))</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>lgd1 <span class="op">=</span> Legend(labels<span class="op">=</span>c(<span class="st">"&lt;0.0001"</span>,<span class="st">"&lt;0.001"</span>,<span class="st">"&lt;0.01"</span>,<span class="st">"&lt;0.05"</span>),pch <span class="op">=</span> c(<span class="st">"**** "</span>,<span class="st">"*** "</span>,<span class="st">"** "</span>,<span class="st">"* "</span>), title <span class="op">=</span> <span class="st">"adj.p"</span>, <span class="bu">type</span> <span class="op">=</span> <span class="st">"points"</span>, </span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>    background <span class="op">=</span> <span class="st">"white"</span>,direction <span class="op">=</span> <span class="st">"horizontal"</span>)</span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>lgd2 <span class="op">=</span> Legend(labels<span class="op">=</span><span class="st">"gene not detected"</span>,legend_gp <span class="op">=</span> gpar(fill <span class="op">=</span> <span class="st">"grey"</span>), title <span class="op">=</span> <span class="st">"NA"</span>)</span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a>pdf(<span class="st">"soc_FvsM_DEG_logFC_heatmap.pdf"</span>,width<span class="op">=</span><span class="dv">6</span>,height<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb33-21"><a href="#cb33-21" aria-hidden="true" tabindex="-1"></a>draw(p,annotation_legend_list <span class="op">=</span> <span class="bu">list</span>(lgd1,lgd2),heatmap_legend_side<span class="op">=</span><span class="st">"bottom"</span>, annotation_legend_side<span class="op">=</span><span class="st">"right"</span>,</span>
<span id="cb33-22"><a href="#cb33-22" aria-hidden="true" tabindex="-1"></a>           legend_grouping <span class="op">=</span> <span class="st">"original"</span>)</span>
<span id="cb33-23"><a href="#cb33-23" aria-hidden="true" tabindex="-1"></a>dev.off()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“The input is a data frame-like object, convert it to a matrix.”
Warning message:
“You defined `cell_fun` for a heatmap with more than 100 rows or
columns, which might be very slow to draw. Consider to use the
vectorized version `layer_fun`.”</code></pre>
</div>
<div class="cell-output cell-output-display">
<strong>png:</strong> 2
</div>
</div>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    const typesetMath = (el) => {
      if (window.MathJax) {
        // MathJax Typeset
        window.MathJax.typeset([el]);
      } else if (window.katex) {
        // KaTeX Render
        var mathElements = el.getElementsByClassName("math");
        var macros = [];
        for (var i = 0; i < mathElements.length; i++) {
          var texText = mathElements[i].firstChild;
          if (mathElements[i].tagName == "SPAN") {
            window.katex.render(texText.data, mathElements[i], {
              displayMode: mathElements[i].classList.contains('display'),
              throwOnError: false,
              macros: macros,
              fleqn: false
            });
          }
        }
      }
    }
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        for (let i = 0; i < 2; i++) {
          container.appendChild(note.children[i].cloneNode(true));
        }
        typesetMath(container);
        return container.innerHTML
      } else {
        typesetMath(note);
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      typesetMath(note);
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>