---
toc: false
---

# ScaleBio run on rat brain tissues from social isolation study
Lin Yang

## Project description
Three replicates for female and male rat treated with social isolation and social grouped respectively are included in the experiment. The treatment was exerted from p41-p55. Then, orbitofrontal cortex are dissected from frozen brain tissues and Hang did the nuclei preparation and library preparation with ScaleBio RNA-seq kit.

A relavant presentation for this project is available at: [presentation](https://uchicago.box.com/s/o20tc5qjuk0ekn20dbln4atk9shx8sb1)

## Data processing steps
1. data preprocess

The library sequencing was ran on Illumina sequencer and demultiplexed by i5 barcode by the sequencing core in UofC.
Four .fastq files are returned from the core: 

* Undetermined_S0_L002_I1_001.fastq.gz: i7 index
* Undetermined_S0_L002_I2_001.fastq.gz: i5 index
* Undetermined_S0_L002_R1_001.fastq.gz: reads1, recording barcode sequences
* Undetermined_S0_L002_R2_001.fastq.gz: reads2, recording cDNA sequences

The data preprocessing are implemented with pipeline from ScaleBio with the command:

```
module load singularity
nextflow run /project/zhuzhuzhang/lyang/software/ScaleRna/ -params-file runParams.yml -profile singularity --outDir /project/zhuzhuzhang/lyang/scaleBio_RNA_rat_brain/out_mRatBN â€”-splitFastq true
```

**inputs:**

[runParams.yml](../script/runParams.yml)

**outputs:**

quality control (/project/zhuzhuzhang/lyang/scaleBio_RNA_rat_brain/out_mRatBN/samples)

cell by feature matrix (/project/zhuzhuzhang/lyang/scaleBio_RNA_rat_brain/out_mRatBN/alignment)

2. cell clustering

[Cell clustering was perform in R with Seurat package](cell_clustering.html)

3. Marker gene indentification in each cluster and cell type annotation

Marker genes in each cluster are identified with `FindAllmarker` function in Seurat. The top 10 genes are saved and manually compared with the database to annotate the cell types for each cluster.

[**script**](cluster_annotation.html)

**outputs:**

[Cluster highlighted in umaps.](https://drive.google.com/file/d/1p-tYEjZUDYtx7mR7NflJWEPTR0LNpI-M/view?usp=drive_link)
[The umap projections of expression marker gene.](https://drive.google.com/file/d/1xW9gSP3w7PrulhY3mcOuhQFS0VB0sfg1/view?usp=drive_link)
[Cell type annotation gene list.](https://docs.google.com/spreadsheets/d/17Y31mBy9B0aljy72LLnAPNpoxCN6zZLhlXBFzCg-awM/edit?usp=drive_link)

4. differential expression analysis between social isolation and social grouped samples

The comparison are based on pseudobulk expression of each cell type in each sample.

[**script**](DEG_analysis_iso_vs_soc.html)

**outputs:**

[DEG counts in each cell type between iso and soc groups](/outputs/DEG_FC0.3_counts_barplot_2.pdf)
[gene lists from the DEG analysis and filtered significantly differentially expressed genes with criteria: adjusted.p<0.05 & |log2FC|>0.3](https://drive.google.com/drive/folders/1vyPnJ4aQlVgrAcMo2c_cSNqdUHIiz79B?usp=drive_link)


5. visualization on the DEGs between social isolation and social grouped comparison

[**script**](iso_vs_soc_DEG_visualization.html)

**outputs:**

[heatmap for log2FC of DEGs in iso&soc comparison](/outputs/isoVSsoc_DEG_logFC_heatmap.pdf)
[Violin plots of DEG expression in iso and soc groups](https://drive.google.com/file/d/12bkUK9wwWnwjBqQUuQmsxsbzI_nVF-8V/view?usp=drive_link)
[umap projecting the expression level of DEGs in each cell](https://drive.google.com/file/d/1XUcQwIZnNIeu8fGhCLbE9HwYXYf-RSao/view?usp=drive_link)

6. differential expression analysis between female and male samples with social isolation and social grouped groups

[**script**](DEG_analysis_F_vs_M.html)

**outputs:**

[gene lists from the DEG analysis and filtered significantly differentially expressed genes with criteria: adjusted.p<0.05 & |log2FC|>0.3](https://drive.google.com/drive/folders/1R0_wEIfob2jwujwN7ijTYtl6ExWU-4Sz?usp=drive_link)



7. visualization on the DEGs between female and male comparison

[**script**](sex_DEG_plot.html)

**outputs:**

heatmaps: 

[heatmap for log2FC of DEGs in F&M comparison](https://drive.google.com/drive/folders/1-MPa-pQrpmmMFsI-UmCmYJUmbA7dZKXz?usp=drive_link)

umaps: 

[umap projecting the expression level of DEGs between F&M comparison in soc group](https://drive.google.com/file/d/1ujv1Oxv-nIDN9WiGDA9KChfhR6LS5nWl/view?usp=drive_link) &

[umap projecting the expression level of DEGs between F&M comparison in iso group](https://drive.google.com/file/d/1vyZy_f4YklpoiqHshb8UcBFwxCYjaVjk/view?usp=drive_link)

violin plot: 

[violin plot of gene expression of DEGs between F&M comparison in soc group](https://drive.google.com/file/d/1GnPFiDTRAboEHK_34HY5XSBzRK9bS01T/view?usp=drive_link)

[violin plot of gene expression of DEGs between F&M comparison in iso group](https://drive.google.com/file/d/1elizpJDO6-6KoptOsUv8xzSR96Oi7R4r/view?usp=drive_link)

8. genomic track visualization 

steps involved:

* [split the integrated .bam files into individual cell based on the barcode information](../script/workflow.sh)

* [To gather the single-cell .bam files to the same folder annotated as the same cell type (not necessarily move to same folder, create softlink works)](../script/run_bam_gather.sh)

* [Merge the .bam files together for each cell type in each sample](/script/samtools_merge.sh)

* [Make .bigwig from the merged .bam file](/script/make_bigwig.sh)

* [make Genome tracks with pygenometrack](/script/soc_FvsM_pygenometrack.sh)

